### ISSUE TITLE
Mobs not standing correctly on slabs and lilypads

### ISSUE THREAD

**Issue by falkseidl** [mc-server/MCServer#2015]
Mobs standing in the height of a full block, if they are walking on blocks that have a smaller height.
It happens, if they walk for their self on this blocks, but also if the player spawns a mob at that place.

Screenshots:
![2015-05-14_10 59 10](https://cloud.githubusercontent.com/assets/5732964/7628857/457f8eb2-fa28-11e4-87e0-225a0ee63f2a.png)

![2015-05-14_10 58 36](https://cloud.githubusercontent.com/assets/5732964/7628843/3029b272-fa28-11e4-996c-a41028a638ab.png)

**Comment by SafwatHalaby** [comment]
The pathfinder has partial blocks awareness yet, it's in the [to-do list](http://forum.mc-server.org/showthread.php?tid=1895) and will be fixed :)

Thanks for pointing out the lilypad issue.

**Comment by madmaxoft** [comment]
I don't think the pathfinder should be partial-block-aware. It should only provide paths through entire blocks.

This is an issue of the physics engine, though, that should really be aware of correct block hitboxes.

**Comment by SafwatHalaby** [comment]
Good point, but in the meanwhile this can be delegated to `monster.cpp`:
"If next waypoint is in the air and there's a slab beneath it, lower it half a block"
"If next waypoint is in the air and there's a lilypad beneath it, lower it 1 block"

**Comment by falkseidl** [comment]
Happens also on trapdoors or snow. So all special blocks like this.

**Comment by SafwatHalaby** [comment]
yes, all partial blocks are like this.

**Comment by SafwatHalaby** [comment]
I tried fixing it, but as @madmaxoft  said, looks like it requires changes to the physics engine. The physics doesn't care about half slabs at all. I completely neutralized the pathfinder and the mobs were still flying when spawned over slabs.

**Comment by worktycho** [comment]
So basicly the physics engine needs to have full bounding box support.

**Comment by SafwatHalaby** [comment]
Doesn't have to be full, can be "cheaty full", give special treatment to exotic blocks.

**Comment by SafwatHalaby** [comment]
Fixing lilypads and snow sounds easy - I'll try to make the physics engine simply ignore flat blocks.

**Comment by worktycho** [comment]
That was an accident. Are there any blocks that aren't either empty or half blocks?

**Comment by worktycho** [comment]
I forgot, steps. What other shapes are there.

**Comment by SafwatHalaby** [comment]
The bed is slightly more than half, but really, we shouldn't care that little precision.

**Comment by worktycho** [comment]
So thats three special cases. We can probably repurpose FullyOccupiesVoxel to store the information.

**Comment by SafwatHalaby** [comment]
That's two. Forget the bed and treat it like a slab! :P

**Comment by worktycho** [comment]
empty, stair and slab, its three.

**Comment by SafwatHalaby** [comment]
The physics engine doesn't look simple and I want to focus on the AI stuff right now. Could anyone fix the physics for me? Ignore lilypads and snow, give special treatment to half blocks.

Regarding stairs: The mobs treat them like single blocks and jump, so it's relatively fine and not urgent.

**Comment by worktycho** [comment]
So a temporary fix would be to special case slabs and empty.

**Comment by SafwatHalaby** [comment]
Yes, and stairs can be dealt with later, because they're good enough. I think the stairs will require some special treatment from the AI side as well (mini jumps), so it's better left for later.

**Comment by SafwatHalaby** [comment]
By the way, paralyzing the mobs for debugging is as simple as placing a `return false` at the first line of `cMonster::TickPathFinding(...)` .

**Comment by worktycho** [comment]
List of places that need fixing:
 - cBoundingBox::CalcLineIntersection
 - cProjectileEntity::OnHitSolidBlock
 - cProjectileTracerCallback::OnNextBlock
 - cEntity::HandlePhysics

I personally don't think I can make the change, because I'm not set up for testing that sort of change.

**Comment by SafwatHalaby** [comment]
I did not dive deep into the code so you know better, but shouldn't the lilypad part be much simpler? I imagined it would be something along the lines of change an `if (is_air_block)...` to `if (is_air_block || is_lily_block)`, that is, treat lily as air and that's it.

**Comment by SafwatHalaby** [comment]
Oh, mobs stand on lily, so it's a bit more involved. (At the very least: treat as air and treat block beneath as solid)

**Comment by madmaxoft** [comment]
Not exactly, the mobs stand on the lilypad itself, it is a 1/16 height solid block. Or even 1/32? I don't remember now.

**Comment by SafwatHalaby** [comment]
We could cheat like that. Who'd care?

**Comment by DarkoGNU** [comment]
It's been like seven years since the last comment on this issue, so I think it deserves a little update. There's an even bigger issue with snow, in my opinion it deserves its own issue...

Client version: 1.12.2
Server OS: Linux
Cuberite Commit id: 900f95c42b9b6ff333cc60640c17fb197025e280

### Expected behavior

1. Mobs stand correctly on blocks that aren't full blocks.
2. Mobs receive fall damage when falling on these blocks.
3.  Mobs trample farmlands (after merging #5401).

### Actual behavior

1. Mobs stand on these blocks as if they were full blocks.
2. Mobs don't receive fall damage when falling on these blocks. Instead, they receive the damage... after walking off them, onto a full block!
3. Mobs don't trample farmlands (after merging #5401).

### Blocks affected

It seems like all blocks that aren't full blocks are affected. Some examples:

- slabs
- farmlands
- enchantment tables
- beds

**Comment by DarkoGNU** [comment]
<work in progress, sorry if I've posted by accident>

It's been like seven years since the last comment on this issue, so I think it deserves a little update. There's an even bigger issue with snow, in my opinion it deserves its own issue...

Client version: 1.12.2
Server OS: Linux
Cuberite Commit id: 900f95c42b9b6ff333cc60640c17fb197025e280

### Expected behavior

1. Mobs stand correctly on blocks that aren't full blocks
2. Mobs receive fall damage when falling on these blocks
3.  Mobs trample farmlands (after merging #5401)

### Actual behavior

1. Mobs stand on these blocks as if they were full blocks. This is true for all nonfull blocks
2. Mobs don't receive fall damage when falling on these blocks. Instead, they receive the damage... after walking off them, onto a full block!

This is true only for blocks for which `cBlockInfo::GetBlockHeight` returns the right height. For example, beds are affected - this method reports their height as `0.5625`. Daylight sensors aren't affected, they're not included in that method - so I assume that the default value (`1`) is returned for them.

3. Mobs don't trample farmlands (after merging #5401)

### Blocks affected

It seems like all blocks that aren't full blocks are affected. Some examples:

- slabs
- farmlands
- enchantment tables
- beds

### Steps to reproduce the behavior

1. Place some affected blocks
2. Somehow place mobs on them. They can just walk onto them, you may drop them, or use a spawn egg. For the visual bug - anything works

If testing for the fall damage bug or farmland trampling, drop the mobs from some height, of course. Better make sure they stand on a solid block before the drop, else they won't take any damage, no matter what - but that's another bug (#5323).

3. If testing for that fall damage bug, check if the tested block is included in `cBlockInfo::GetBlockHeight`

### Videos

The fall damage bug is a little bit funny, I think is deserves its own video. [Click here](https://www.youtube.com/watch?v=TM8dengrzRo)

### Screenshots

![image](https://user-images.githubusercontent.com/42816979/164119404-54232afd-9c45-4222-b594-1b9a563cfa43.png)

![image](https://user-images.githubusercontent.com/42816979/164119563-8da39f2e-e506-45ab-9592-0c7fec6356e9.png)
